### PR TITLE
Add support for FHIR extensions.

### LINKED ISSUE #163: Extension content
This change will add support for the `extension` FHIRPath function (see [Additional functions](https://hl7.org/fhir/R4/fhirpath.html#functions)).

As part of this, changes will be made to persist extension content upon import to enable its query using the `extension` function.

See also: [Extension content](https://pathling.csiro.au/docs/roadmap.html#extension-content)

### PR DESCRIPTION
**Pull request by piotrszul**
Resolves #163

### CODE REVIEW

**Review by johngrimes — CHANGES_REQUESTED** [review]
Looks great, thanks for your work!

I've added a few questions and suggestions for you to take a look at.

File: encoders/src/main/scala/au/csiro/pathling/encoders/datatypes/R4DataTypeMappings.scala [hunk 1]
```diff
@@ -106,7 +132,12 @@ class R4DataTypeMappings extends DataTypeMappings {
     val skipContains = definition.getImplementingClass == classOf[ValueSet.ValueSetExpansionContainsComponent] &&
       child.getElementName == "contains"
 
-    skipRecursiveReference || skipContains
+    // TODO: This is due to a bug in happi RuntimeChildExtension.getChildByName() implementation
+    // which fails on assestion because the name of the child should be "modifierExtensionExtension"
+    // not "extensionExtension"
+    val skipModifierExtension = child.getElementName.equals("modifierExtension")
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/main/scala/au/csiro/pathling/encoders/datatypes/R4DataTypeMappings.scala:138]: Reported as: https://github.com/hapifhir/hapi-fhir/issues/3414
Updated the comment as well.

File: encoders/src/main/scala/au/csiro/pathling/encoders2/SerializerBuilder2.scala [hunk 2]
```diff
@@ -192,16 +245,38 @@ private[encoders2] object SerializerBuilderProcessor {
...
+    def flattenBase(obj: Base): List[(Int, java.util.List[Extension])] = {
+
+      val childrenExts = resource.children()
+        .map(p => obj.getProperty(p.getName.hashCode, p.getName, false))
+        .filter(_ != null)
+        .flatMap(_.flatMap(flattenBase))
+        .toList
+
+      obj match {
+        // TODO: fix the type cast
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/main/scala/au/csiro/pathling/encoders2/SerializerBuilder2.scala:260]: None

File: encoders/src/test/java/au/csiro/pathling/encoders/ResourceEncodingTest.java [hunk 3]
```diff
@@ -29,10 +29,11 @@
 public class ResourceEncodingTest {
 
   private final FhirContext fhirContext = FhirContext.forR4();
-  private final SchemaConverter1 converter = new SchemaConverter1(fhirContext,
-      new R4DataTypeMappings(), 0);
+  private final SchemaConverter2 converter = new SchemaConverter2(fhirContext,
+      new R4DataTypeMappings(), EncoderConfig.apply(0, true));
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/test/java/au/csiro/pathling/encoders/ResourceEncodingTest.java:33]: It should be SchemaConverter2.
It's designed to test the current default version (v2).

File: fhir-server/src/main/scala/au/csiro/pathling/sql/PathlingExpressions.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.spark.sql.types.{DataType, StructType}
+
+
+/**
+ * An expression that removes all fields starting with '_' underscores from struct values.
+ * Other types are not affected.
+ *
+ * @param child the child expresion.
+ */
+case class Normalize(child: Expression) extends UnaryExpression with CodegenFallback with NullIntolerant {
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply fhir-server/src/main/scala/au/csiro/pathling/sql/PathlingExpressions.scala:21]: Changed to `PruneSyntheticFields`

File: encoders/src/test/java/au/csiro/pathling/encoders/TestData.java [hunk 5]
```diff
@@ -355,4 +356,59 @@ public static QuestionnaireResponse newNestedQuestionnaireResponse(final int max
...
+    //    - url: uuid:nested
+    //    - value: nested
+    // - identifier
+    //     - id: uuid:identifier1
+    //     - extension:
+    //       - url: uuid:ext10
+    //       - value: ext10
+    //     - extension:
+    //        - url: uuid:ext11
+    //        - value: 11
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/test/java/au/csiro/pathling/encoders/TestData.java:382]: No, at the moment extending of primitive elements is not implemented (and thus we would not be able use this specific extension type as it extends primitives dateTime/time).
I was thinking to address these more exotic cases as a seperate issue once we understand the impact of the current implementation for composite types on performance and validate in in general.

File: encoders/src/main/scala/au/csiro/pathling/encoders2/SchemaTraversal.scala [hunk 6]
```diff
@@ -10,188 +10,404 @@
...
+   */
+  def visitResource(resourceCtx: ResourceCtx[DT, SF]): DT = {
+    resourceCtx.proceed(this)
+  }
+}
+
+object SchemaVisitor {
+  def traverseResource[DT, SF](resourceDefinition: RuntimeResourceDefinition, visitor: SchemaVisitor[DT, SF]): DT = {
+    // TODO: NOT SURE where is the best place to put it
+    // Not if it will be needed anymore if I can do recursive contexts
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/main/scala/au/csiro/pathling/encoders2/SchemaTraversal.scala:109]: None.

File: encoders/src/main/scala/au/csiro/pathling/encoders2/DeserializerBuilder2.scala [hunk 7]
```diff
@@ -145,6 +160,26 @@ private[encoders2] class DeserializerBuilderProcessor(val path: Option[Expressio
...
+  private def deserializeExtensions(definition: BaseRuntimeElementCompositeDefinition[_], beanExpression: Expression): Expression = {
+    val beanWithFid = RegisterFid(beanExpression, addToPath(ExtensionSupport.FID_FIELD_NAME))
+    definition match {
+      case _: RuntimeResourceDefinition =>
+        val deserializeExtension = UnresolvedCatalystToExternalMap(
+          addToPath("_extension"),
+          identity,
+          /// TODO: Fix the encoding context creation
+          // This function somehow managed to be called ouside the scope of the current traversal
+          // Which may actually be OK, but s
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/main/scala/au/csiro/pathling/encoders2/DeserializerBuilder2.scala:172]: Updated the comment.

File: encoders/src/main/scala/au/csiro/pathling/encoders/EncoderContext.scala [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * Default implementation of [[EncoderSettings]].
+ *
+ * @param maxNestingLevel    @see  [[EncoderSettings.maxNestingLevel]]
+ * @param supportsExtensions @see  [[EncoderSettings.supportsExtensions]]
+ * @param generateFid        @see [[EncoderSettings.generateFid]]
+ */
+case class EncoderConfig(override val maxNestingLevel: Int,
+                         override val supportsExtensions: Boolean,
+                         override val generateFid: Boolean) extends EncoderSettings {
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/main/scala/au/csiro/pathling/encoders/EncoderContext.scala:77]: Maybe but I am not sure about it. 
This config is more 'internal' and at least until we get rid of V1 encoders have options that should be available for FhirEncoders.
Perhaps we could revisit it when we remove V1 encoders.

File: encoders/src/main/java/au/csiro/pathling/encoders/FhirEncoders.java [hunk 9]
```diff
@@ -295,14 +307,17 @@ public int hashCode() {
   public static class Builder {
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/main/java/au/csiro/pathling/encoders/FhirEncoders.java:307]: Updated.

File: fhir-server/src/test/java/au/csiro/pathling/fhirpath/parser/ParserTest.java [hunk 10]
```diff
@@ -530,4 +520,93 @@ void parserErrorThrows() {
...
+
+    // Create a mock terminology client.
+    when(terminologyService.translate(any(), any(), anyBoolean(), any()))
+        .thenReturn(returnedConceptTranslator);
+
+    assertThatResultOf(ResourceType.CONDITION,
+        "code.coding.translate('http://snomed.info/sct?fhir_cm=900000000000526001', false, 'equivalent').extension('uuid:any').url")
+        .selectOrderedResult()
+        .hasRows(spark, "responses/ParserTest/testExtensionFunctionOnTranslateResult.csv");
+  }
```
[the remark this answers is not in the document]
Comment (johngrimes) [reply fhir-server/src/test/java/au/csiro/pathling/fhirpath/parser/ParserTest.java:611]: This has now been fixed on `release/5.0.0`, and I have merged it in and confirmed that it is working.

File: encoders/src/test/java/au/csiro/pathling/encoders/TestData.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (johngrimes) [reply encoders/src/test/java/au/csiro/pathling/encoders/TestData.java:382]: I have created a new issue for this: #437

File: encoders/src/main/scala/au/csiro/pathling/encoders/EncoderContext.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (johngrimes) [reply encoders/src/main/scala/au/csiro/pathling/encoders/EncoderContext.scala:77]: I have added an issue relating to the removal of the v1 encoders: #438

**Review by johngrimes — APPROVED** [review]

**Pull request merged**